File: BinarySearch/GFG/Easy/NthRoot/NthRoot.cpp
// https://www.geeksforgeeks.org/problems/find-nth-root-of-m5843/1
#include <bits/stdc++.h>
using namespace std;
long long int findPower(long long int x, long long int y, int m)
{

    long long int ans = 1;
    while (y > 0)
    {
        if (y & 1)
        {
            ans *= x;
            if (ans > m)
                return INT_MAX;
            y -= 1;
        }
        else
        {
            x = x * x;
            if (x > m)
                return INT_MAX;
            y /= 2;
        }
    }
    return ans > 0 ? ans : INT_MAX;
}
int NthRoot(int n, int m)
{
    // Code here.
    long long int l = 1, u = m;
    while (l <= u)
    {
        long long int mid = u - (u - l) / 2;
        long long int ans = findPower(mid, n, m);
        if (ans == m)
        {
            return mid;
        }
        else if (ans < m)
        {
            l = mid + 1;
        }
        else
        {
            u = mid - 1;
        }
    }
    return -1;
}
int main(void)
{

    return 0;
}

File: BinarySearch/GFG/Easy/NthRoot/NthRoot.java
// https://www.geeksforgeeks.org/problems/find-nth-root-of-m5843/1
public class NthRoot {
    int findPower(long x, int y, int m) {
        long ans = 1;
        while (y > 0) {
            if ((y & 1) == 1) {
                ans = ans * x;
                if (ans > m) {
                    return Integer.MAX_VALUE;
                }
                y = y - 1;
            } else {
                x *= x;
                if (x > m) {
                    return Integer.MAX_VALUE;
                }
                y /= 2;
            }
        }
        return (int) ans;
    }

    public int NthRoot(int n, int m) {
        // code here
        int l = 1, u = m;
        while (l <= u) {
            int mid = u - (u - l) / 2;
            int ans = findPower(mid, n, m);
            if (ans == m) {
                return mid;
            } else if (ans < m) {
                l = mid + 1;
            } else {
                u = mid - 1;
            }
        }
        return -1;
    }
}
